### PR TITLE
CLI: extension registry commands

### PR DESCRIPTION
**Pull request by stefannica**
Aside from implementing the extension registry CLI commands, this commit
also makes some other relate changes:

* add extension query/list operation to REST API, domain and extension
manager
* update identifier fields used in REST API extension objects to allow
empty ID values to be supplied
* allow empty extensions and services in the extension query response
returned by te extension registry. This filter is redundand, as it's
also being done by the extension manager

Implements: https://github.com/fuseml/fuseml/issues/74

CLI examples:

```
fuseml extension list
+--------------+--------------+---------+------+-------------------------------+-----------------------------------------+----------------------------+
| ID           | PRODUCT      | VERSION | ZONE | SERVICES                      | ENDPOINTS                               | CREDENTIALS                |
+--------------+--------------+---------+------+-------------------------------+-----------------------------------------+----------------------------+
| cert-manager | cert-manager | 1.4.0   |      |                               |                                         |                            |
| kfserving    | kfserving    | 0.6.0   |      | [ API ]                       | [ API ]                                 |                            |
|              |              |         |      | resource: kfserving-api       | internal: http://kubernetes.default.svc |                            |
|              |              |         |      | category: prediction-serving  |                                         |                            |
|              |              |         |      |                               |                                         |                            |
|              |              |         |      | [ kfserving-ui-j6pprdrc ]     |                                         |                            |
|              |              |         |      | resource: kfserving-ui        |                                         |                            |
|              |              |         |      | category: UI                  |                                         |                            |
|              |              |         |      |                               |                                         |                            |
|              |              |         |      |                               |                                         |                            |
| knative      | knative      | 0.24.0  |      |                               |                                         |                            |
| mlflow       | mlflow       | 1.19.0  |      | [ mlflow-tracking ]           | [ mlflow-tracking ]                     | [ mlflow-store ]           |
|              |              |         |      | resource: mlflow-tracking     | internal: http://mlflow                 | default-s3-account: global |
|              |              |         |      | category: experiment-tracking |                                         |                            |
|              |              |         |      |                               | [ mlflow-store ]                        |                            |
|              |              |         |      | [ mlflow-store ]              | internal: http://mlflow-minio:9000      |                            |
|              |              |         |      | resource: s3                  |                                         |                            |
|              |              |         |      | category: model-store         |                                         |                            |
|              |              |         |      |                               |                                         |                            |
|              |              |         |      |                               |                                         |                            |
+--------------+--------------+---------+------+-------------------------------+-----------------------------------------+----------------------------+
```

```
fuseml extension get mlflow
ID:            mlflow
Registered:    2021-09-03T06:35:41Z
Description:   MLFlow is an open source platform specialized in tracking ML experiments, and packaging and deploying ML models.

Product:   mlflow
Version:   1.19.0
⛩  Services

 ∙ ID:  mlflow-store
   Description:               MLFlow minio S3 storage back-end
   Resource:                  s3
   Category:                  model-store
   Authentication required:   false
   ⛩  Endpoints

    ∙ URL:  http://mlflow-minio:9000
      Type:   internal
      Configuration:
       ∙ MLFLOW_S3_ENDPOINT_URL = http://mlflow-minio:9000
   ⛩  Credentials

    ∙ ID:  default-s3-account
      Scope:   global
      Configuration:
      ∙ AWS_KEY_ID = <hidden>
      ∙ AWS_SECRET_ACCESS_KEY = <hidden>
 ∙ ID:  mlflow-tracking
   Description:               MLFlow experiment tracking service API and UI
   Resource:                  mlflow-tracking
   Category:                  experiment-tracking
   Authentication required:   false
   ⛩  Endpoints

    ∙ URL:  http://mlflow
      Type:   internal
      Configuration:
       ∙ MLFLOW_TRACKING_URI = http://mlflow
```

```
fuseml extension service list mlflow
+-----------------+-----------------+---------------------+---------------+
| ID              | RESOURCE        | CATEGORY            | CONFIGURATION |
+-----------------+-----------------+---------------------+---------------+
| mlflow-store    | s3              | model-store         |               |
| mlflow-tracking | mlflow-tracking | experiment-tracking |               |
+-----------------+-----------------+---------------------+---------------+
```

```
fuseml extension endpoint list mlflow mlflow-store
+--------------------------+----------+--------------------------------------------------+
| URL                      | TYPE     | CONFIGURATION                                    |
+--------------------------+----------+--------------------------------------------------+
| http://mlflow-minio:9000 | internal | MLFLOW_S3_ENDPOINT_URL: http://mlflow-minio:9000 |
+--------------------------+----------+--------------------------------------------------+
```

```
fuseml extension credentials list mlflow mlflow-store
+--------------------+--------+----------+-------+---------------------------------+
| ID                 | SCOPE  | PROJECTS | USERS | CONFIGURATION                   |
+--------------------+--------+----------+-------+---------------------------------+
| default-s3-account | global |          |       | AWS_KEY_ID: <hidden>            |
|                    |        |          |       | AWS_SECRET_ACCESS_KEY: <hidden> |
+--------------------+--------+----------+-------+---------------------------------+
```

### CODE REVIEW

**Comment by jsuchome** [comment]
As you are already touching the example files, could you fix that AWS_KEY value too?

File: pkg/cli/common/util.go [hunk 1]
```diff
@@ -102,3 +109,21 @@ func WriteConfigFile() error {
 	}
 	return nil
 }
+
+// StringInSlice verifies if a string slice contains a string value
+func StringInSlice(s string, slice []string) bool {
```
Comment (jsuchome) [inline pkg/cli/common/util.go:114]: This is already used in `pkg/core/extension_store.go`. Could you move the method to common place which could be imported from anywhere in core code?

Comment (jsuchome) [reply pkg/cli/common/util.go:114]: BTW there's also `contains` method (same thing) declared at 3 different places in core...

File: pkg/cli/extension/delete.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+func newExtensionDeleteOptions(o *common.GlobalOptions) (res *extensionDeleteOptions) {
+	return &extensionDeleteOptions{global: o}
+}
+
+func newSubCmdExtensionDelete(gOpt *common.GlobalOptions) *cobra.Command {
+	o := newExtensionDeleteOptions(gOpt)
+	cmd := &cobra.Command{
+		Use:   "delete {EXTENSION_ID}",
+		Short: "Deletes a extension",
+		Long:  `Delete a extension from the FuseML extension registry, along with all services, endpoints and credentials.`,
```
Comment (jsuchome) [inline pkg/cli/extension/delete.go:26]: _an extension_ (same for line above)

File: pkg/cli/extension/endpoint/delete.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+func newEndpointDeleteOptions(o *common.GlobalOptions) (res *endpointDeleteOptions) {
+	return &endpointDeleteOptions{global: o}
+}
+
+func newSubCmdEndpointDelete(gOpt *common.GlobalOptions) *cobra.Command {
+	o := newEndpointDeleteOptions(gOpt)
+	cmd := &cobra.Command{
+		Use:   "delete {EXTENSION_ID} {SERVICE_ID} {ENDPOINT_URL}",
+		Short: "Deletes a endpoint from an extension service",
+		Long:  `Delete a endpoint from an extension service registered with the FuseML extension registry.`,
```
Comment (jsuchome) [inline pkg/cli/extension/endpoint/delete.go:26]: _an endpoint_

File: pkg/cli/extension/endpoint/update.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+func newEndpointUpdateOptions(o *common.GlobalOptions) *endpointUpdateOptions {
+	return &endpointUpdateOptions{global: o}
+}
+
+func newSubCmdEndpointUpdate(gOpt *common.GlobalOptions) *cobra.Command {
+	o := newEndpointUpdateOptions(gOpt)
+	cmd := &cobra.Command{
+		Use:   `update [--type {internal|external}] [-c|--configuration KEY:VALUE]... {EXTENSION_ID} {SERVICE_ID} {ENDPOINT_URL}`,
+		Short: "Update the attributes of an existing FuseML extension endpoint",
```
Comment (jsuchome) [inline pkg/cli/extension/endpoint/update.go:29]: While we're are in grammar phase of a review ... isn't it rather _of existing FuseML extension endpoint_ ? Without the article here, but I'm not sure....

File: pkg/cli/extension/get.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	"text/template"
+
+	"github.com/spf13/cobra"
+	"github.com/tektoncd/cli/pkg/formatted"
+
+	"github.com/fuseml/fuseml-core/pkg/cli/client"
+	"github.com/fuseml/fuseml-core/pkg/cli/common"
+)
+
+const extensionGetTemplate = `{{decorate "bold" "ID"}}:	{{ .ID }}
```
Comment (jsuchome) [inline pkg/cli/extension/get.go:16]: Wow, you really wrote a go template ... I'm not gonna read through this, so I'll just ask: is this a necessary way of doing it?

File: pkg/cli/extension/get.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	res := &extensionGetOptions{global: o}
+	res.format = common.NewSingleValueFormattingOptions()
+	return res
+}
+
+func newSubCmdExtensionGet(gOpt *common.GlobalOptions) *cobra.Command {
+	o := newExtensionGetOptions(gOpt)
+	cmd := &cobra.Command{
+		Use:   `get {EXTENSION_ID}`,
+		Short: "Get a extension",
```
Comment (jsuchome) [inline pkg/cli/extension/get.go:109]: `an`

File: pkg/cli/extension/list.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	res.format = common.NewFormattingOptions(
+		[]string{"ID", "Product", "Version", "Zone", "Services", "Endpoints", "Credentials"},
+		[]table.SortBy{{Name: "ID", Mode: table.Asc}},
+		common.OutputFormatters{"Services": formatServices, "Endpoints": formatEndpoints, "Credentials": formatCredentials},
+	)
+
+	return
+}
+
+func derefString(s *string, defaultValue ...string) string {
```
Comment (jsuchome) [inline pkg/cli/extension/list.go:34]: couldn't you use `derefString` that you've already defined somewhere else?

File: pkg/cli/extension/endpoint/update.go (hunk 4 above)
Comment (stefannica) [reply pkg/cli/extension/endpoint/update.go:29]: it's either "of an existing endpoint" or "of existing endpoints", but both are correct

File: pkg/cli/extension/get.go (hunk 5 above)
Comment (stefannica) [reply pkg/cli/extension/get.go:16]: I wouldn't say it's necessary, but it does makes it easier to read :)
we do the same for `workflow get`, I just borrowed the idea

File: pkg/cli/common/util.go (hunk 1 above)
Comment (stefannica) [reply pkg/cli/common/util.go]: took a lot of changes, but I moved it to a common util package

File: pkg/cli/extension/delete.go (hunk 2 above)
Comment (stefannica) [reply pkg/cli/extension/delete.go]: fixed

File: pkg/cli/extension/endpoint/delete.go (hunk 3 above)
Comment (stefannica) [reply pkg/cli/extension/endpoint/delete.go]: fixed

File: pkg/cli/extension/get.go (hunk 6 above)
Comment (stefannica) [reply pkg/cli/extension/get.go]: fixed

File: pkg/cli/extension/list.go (hunk 7 above)
Comment (stefannica) [reply pkg/cli/extension/list.go]: took even more changes, but I moved all ref/deref functions to the common util package

**Comment by stefannica** [comment]
> As you are already touching the example files, could you fix that AWS_KEY value too?

done

File: pkg/cli/extension/get.go (hunk 5 above)
Comment (jsuchome) [reply pkg/cli/extension/get.go:17]: _easier to read_ in combination of _golang templates_? That's something...

File: pkg/cli/common/util.go (hunk 1 above)
Comment (jsuchome) [reply pkg/cli/common/util.go]: great!

File: pkg/cli/extension/register.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func newExtensionRegisterOptions(o *common.GlobalOptions) *extensionRegisterOptions {
+	return &extensionRegisterOptions{global: o}
+}
+
+func newSubCmdExtensionRegister(gOpt *common.GlobalOptions) *cobra.Command {
+	o := newExtensionRegisterOptions(gOpt)
+	cmd := &cobra.Command{
+		Use:   `register {WORKFLOW_FILE}`,
```
Comment (jsuchome) [inline pkg/cli/extension/register.go:24]: It's not a `WORKFLOW_FILE`

Comment (stefannica) [reply pkg/cli/extension/register.go]: good catch, fixed

**Review by jsuchome — APPROVED** [review]

**Pull request merged**